org.osgi.test.common/src/main/java/org/osgi/test/common/install/InstallBundle.java

```diff
@@ -18,21 +18,294 @@
+				oBundleSymbolicName = Optional.empty();
+				oBundleVersion = Optional.empty();
+				fullPath = parts[0];
+			} else if (parts.length == 2) {
+				oBundleSymbolicName = Optional.of(parts[0]);
+				oBundleVersion = Optional.empty();
+				fullPath = parts[1];
+			} else if (parts.length == 3) {
+				oBundleSymbolicName = Optional.of(parts[0]);
+				oBundleVersion = Optional.of(parts[1]);
+				fullPath = parts[2];
+			} else {
+				throw new IllegalArgumentException("Wrong pattern :" + spec);
+			}
+
+			int indexOfFileBegin = fullPath.lastIndexOf('/') + 1;
+
+			String file = fullPath.substring(indexOfFileBegin);
+
+			if (file.isEmpty()) {
+				throw new IllegalArgumentException(String.format("could not find a filename : ", spec));
+
+			}
+			Optional<String> oPath = Optional.of(fullPath.substring(0, indexOfFileBegin));
+
+			return EmbeddedLocation.of(bundleContext, oBundleSymbolicName, oBundleVersion, oPath, file);
+
+		}
+
+		/**
+		 * Creates an EmbeddedLocation by the pure parameters. existence of File
+		 * not tested at creation time.
+		 *
+		 * @param bundleSymbolicName - symbolicName of the bundle that embedded
+		 *            the file must not be null
+		 * @param bundleVersion - version of the bundle that embedded the file -
+		 *            must not be null
+		 * @param path - the path in the jar - must not be null
+		 * @param file - the filename must not be null
+		 * @return EmbeddedLocation
+		 */
+		public static EmbeddedLocation of(String bundleSymbolicName, Version bundleVersion, String path, String file) {
+			return new EmbeddedLocation(bundleSymbolicName, bundleVersion, path, file);
+		}
+
+		private String	bundleSymbolicName;
+
+		private Version	bundleVersion;
+
+		private String	file;
+		private String	path;
+
+		private EmbeddedLocation() {
+			// no access here.
+		}
+
+		private EmbeddedLocation(String bundleSymbolicName, Version bundleVersion, String path, String file) {
+			if (bundleSymbolicName == null) {
+				throw new IllegalArgumentException("bundleSymbolicName must not be null");
+			}
+			if (bundleVersion == null) {
+				throw new IllegalArgumentException("bundleVersion must not be null");
+			}
+			if (path == null) {
+				throw new IllegalArgumentException("path must not be null");
+			}
+			if (file == null) {
+				throw new IllegalArgumentException("file must not be null");
+			}
+			this.bundleSymbolicName = bundleSymbolicName;
+			this.bundleVersion = bundleVersion;
+
+			path = path.startsWith("/") ? path.substring(1) : path;
+			path = path.endsWith("/") ? path.substring(0, path.length() - 1) : path;
+			this.path = path;
+
+			if (file.startsWith("/")) {
+				file = file.substring(1);
+			}
+			this.file = file;
+		}
+
+		@Override
+		public boolean equals(Object obj) {
+			if (this == obj)
+				return true;
+			if (obj == null)
+				return false;
+			if (getClass() != obj.getClass())
+				return false;
+			EmbeddedLocation other = (EmbeddedLocation) obj;
+			if (bundleSymbolicName == null) {
```

Comment:
This is all a lot easier if you use `Objects.equals`.

```java
return Objects.equals(bundleSymbolicName, other.bundleSymbolicName) && 
  Objects.equals(bundleVersion, other.bundleVersion) && 
  Objects.equals(file, other.file) && 
  Objects.equals(path, other.path);

Reply:
now use Objects methods